### PR TITLE
Optimize AOL adapter and heavily reduce the size

### PR DESCRIPTION
**Pull request by marian-r**
## Type of change

- [ ] Bugfix
- [x] Feature
- [ ] New bidder adapter
- [x] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other

## Description of change

- Refactoring of the AOL adapter, optimizing the logic.
- Removing additional request for a JavaScript library, thus improving performance, reducing the delay and bandwidth usage.
- AOL adapter now also supports deals.
- Adding unit tests (100% code coverage).
- Adding `Number.isInterger()` polyfill.

## Other information

AOL adapter is now maintained by @aol/header-bidding team.

### CODE REVIEW

**Comment by CLKeenan** [comment]
This is great news!! Really looking forward to this.

**Comment by marian-r** [comment]
I think the failed build was not caused by the changes introduced in this PR. It might be related to #642.

**Comment by nedstankus** [comment]
I'm seeing issues with the `window=window` assignment that's in `loaders/delimiterLoader.js`.  You can see this on the example pages; e.g., http://localhost:9999/integrationExamples/gpt/pbjs_example_gpt.html

screen shot 2016-09-27 at 3 00 55 pm

File: src/adapters/aol.js [hunk 1]
```diff
@@ -1,183 +1,137 @@
...
-    bids = params.bids;
-    if (!bids || !bids.length) return;
-    adloader.loadScript(ADTECH_URI, _reqBids, true);
+    utils._each(params.bids, function (bid) {
+      const pubapiUrl = _buildPubapiUrl(bid);
+
+      ajax(pubapiUrl, (response) => {
+        if (!response && response.length <= 0) {
+          utils.logError('Empty bid response', BIDDER_CODE, bid);
+          return;
```
Comment (marian-r) [inline src/adapters/aol.js]: Thanks for the suggestion. This is now fixed.

**Comment by marian-r** [comment]
@nedstankus that is not part of this PR. You are not testing the correct git branch. Please use `contrib/optimized-aol-adapter` branch.

**Comment by nedstankus** [comment]
@marian-r You're correct--`contrib/optimized-aol-adapter` does not have that issue.  I was on `master`.

**Comment by marian-r** [comment]
Are there still any issues/concerns regarding this PR? It has been marked as 'needs review' so do you have an estimate of when the PR could be reviewed?

**Comment by protonate** [comment]
I should be able to review today or tomorrow, thanks for your patience.

**Comment by marian-r** [comment]
These update of AOL adapter also changes the way how gross to net price calculations are handled for AOL bidder. Would it be possible to include a note about this change in the release notes once this is released? I would suggest something like this:

_If you're using AOL as a bidder, please contact your Marketplace account team for a quick config fix (not doing gross to net config client side) as it's now done server side for AOL._

And do you have any estimate of when this could be released? Is it possible to include this in the 0.14.0 milestone?

**Comment by mkendall07** [comment]
@marian-r 
Thanks for the info. Yes we can include that info in the release notes. I've added to 0.14.0 milestone

File: src/adapters/aol.js [hunk 2]
```diff
@@ -1,183 +1,141 @@
...
-   */
-  function _addErrorBid(response, context) {
-    var bid = bidsMap[context.alias];
+  function _addBidResponse(bid, response) {
+    var bidData;
 
-    if (!bid) {
-      utils.logError('mismatched bid: ' + context.placement, ADTECH_BIDDER_NAME, context);
+    try {
+      bidData = response.seatbid[0].bid[0];
```
Comment (protonate) [inline src/adapters/aol.js]: An empty bid or "no bid" response causes this to fail which throws an error, though a response with an empty `seatbid` array appears to be common in my testing, so should be handled as other than Error.

File: src/adapters/aol.js [hunk 3]
```diff
@@ -1,183 +1,141 @@
...
+    if (bidData.ext && bidData.ext.pixels) {
+      ad += bidData.ext.pixels;
     }
 
-    // get the bids
-    utils._each(bids, function (bid) {
-      var bidreq = _mapUnit(bid);
-      window.ADTECH.loadAd(bidreq);
-    });
+    var bidResponse = bidfactory.createBid(1);
```
Comment (protonate) [inline src/adapters/aol.js]: Consider setting the bid response ID to match the bid request ID as described here: https://github.com/prebid/Prebid.js/issues/509. 

Doing so will position the adapter to take advantage of Prebid Roadmap features such as secure iframe support, improved analytics, Prebid Auctions and so on.

File: src/adapters/aol.js [hunk 4]
```diff
@@ -1,183 +1,141 @@
...
+      alias: params.alias || utils.getUniqueIdentifierStr(),
+      bidfloor: (typeof params.bidFloor !== 'undefined') ?
+        `;bidfloor=${params.bidFloor.toString()}` : '',
+      misc: new Date().getTime() // cache busting
+    });
+  }
 
-    // add it to the bid manager
+  function _addErrorBidResponse(bid, response = {}) {
+    var bidResponse = bidfactory.createBid(2);
```
Comment (protonate) [inline src/adapters/aol.js]: Consider setting the bid response ID to match the bid request ID as described here: #509.

Doing so will position the adapter to take advantage of Prebid Roadmap features such as secure iframe support, improved analytics, Prebid Auctions and so on.

Comment (marian-r) [inline src/adapters/aol.js]: Thanks for this suggestion. Done in 38bd3aece22ff2be627bdac53aec0de2bc7f8426.

File: src/adapters/aol.js (hunk 3 above)
Comment (marian-r) [inline src/adapters/aol.js]: Thanks for this suggestion. Done in 38bd3aece22ff2be627bdac53aec0de2bc7f8426.

**Comment by marian-r** [comment]
All issues have been addressed.

File: src/adapters/aol.js [hunk 5]
```diff
@@ -1,183 +1,140 @@
...
+    const params = bid.params;
+    const serverParam = params.server;
+    let regionParam = params.region || 'us';
+    let server;
 
-    cpm = response.getCPM();
-    if (cpm === null || isNaN(cpm)) {
-      return _addErrorBid(response, context);
+    if (!SERVER_MAP.hasOwnProperty(regionParam)) {
+      console.warn(`Unknown region '${regionParam}' for AOL bidder.`);
```
Comment (mkendall07) [inline src/adapters/aol.js]: `console.warn` doesn't play nice in IE 9. Please use `utils.logWarn` instead for safety.

**Comment by mkendall07** [comment]
@marian-r 
Thanks for updating the adapter and adding tests. My last request for you is if you could provide us with test params that return bids for use in our integration testing. Thank you!

File: src/adapters/aol.js (hunk 5 above)
Comment (mkendall07) [inline src/adapters/aol.js]: fair enough I suppose. Thanks

**Comment by marian-r** [comment]
@mkendall07 thanks for merging. I'm just following with the testing params:

This is for 728x90 placements:
```
{
  bidder: 'aol',
  params: {
    network: '9599.1',
    placement: 3675022 // 728x90
  }
}
```

This is for 300x250 placements:
```
{
  bidder: 'aol',
  params: {
    network: '9599.1',
    placement: 3675026 // 300x250
  }
}
```

Please let me know whether they work.

**Pull request merged**